### PR TITLE
toolchain: add support Intel oneApi toolchain based on llvm

### PR DESCRIPTION
**Pull request by nashif**
This adds preliminary support for the oneApi toolchain for X86. The
toolchain is available from here:

   https://software.intel.com/content/www/us/en/develop/tools/oneapi/base-toolkit/download.html

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

### CODE REVIEW

**Comment by nashif** [comment]
@xfl888 FYI

**Comment by ceolin** [comment]
why not simply:
```
include(${ZEPHYR_BASE}/cmake/compiler/clang/compiler_flags.cmake)
```

It seems the changes in this file are the same in clang.

**Comment by nashif** [comment]
@tejlmand  can you please review this?

File: cmake/linker/lld/target.cmake [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    # for Makefile generators and will be ignored by other generators.
+    set(linker_script_dep IMPLICIT_DEPENDS C ${LINKER_SCRIPT})
+  elseif(CMAKE_GENERATOR STREQUAL "Ninja")
+    # Using DEPFILE with other generators than Ninja is an error.
+    set(linker_script_dep DEPFILE ${PROJECT_BINARY_DIR}/${linker_script_gen}.dep)
+  else()
+    # TODO: How would the linker script dependencies work for non-linker
+    # script generators.
+    message(STATUS "Warning; this generator is not well supported. The
+  Linker script may not be regenerated when it should.")
```
Comment (dcpleung) [inline cmake/linker/lld/target.cmake:24]: alignment?

**Review by dcpleung — APPROVED** [review]

File: cmake/bintools/oneApi/target.cmake [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+set(find_program_clang_args )
+set(find_program_binutils_args )
+
+find_program(CMAKE_AR      llvm-ar      ${find_program_clang_args}   )
+find_program(CMAKE_NM      llvm-nm      ${find_program_clang_args}   )
+find_program(CMAKE_OBJDUMP llvm-objdump ${find_program_clang_args}   )
+find_program(CMAKE_RANLIB  llvm-ranlib  ${find_program_clang_args}   )
+find_program(CMAKE_OBJCOPY objcopy      ${find_program_binutils_args})
+find_program(CMAKE_READELF readelf      ${find_program_binutils_args})
+
```
Comment (tejlmand) [inline cmake/bintools/oneApi/target.cmake:12]: Interresting observation, when I run CMake multiple times I notice the following:
- First time: `CMAKE_STRIP` is undefined, thus running ninja result in: `&& --strip-all zephyr.elf -ozephyr.strip` being executed leading to following error:
  `/bin/sh: 1: --strip-all: not found`
- Second run: Re-running CMake using: `cmake .` gets me this as `CMAKE_STRIP` value: `CMAKE_STRIP-NOTFOUND`
- Third run: Re-running CMake once more: `cmake .` gets me this as `CMAKE_STRIP` value: `/usr/bin/strip`

**Review by tejlmand — CHANGES_REQUESTED** [review]
Initial review and testing.

File: cmake/bintools/oneApi/target.cmake [hunk 3]
```diff
@@ -0,0 +1,17 @@
+# SPDX-License-Identifier: Apache-2.0
+
+set(find_program_clang_args )
+set(find_program_binutils_args )
+
+find_program(CMAKE_AR      llvm-ar      ${find_program_clang_args}   )
+find_program(CMAKE_NM      llvm-nm      ${find_program_clang_args}   )
+find_program(CMAKE_OBJDUMP llvm-objdump ${find_program_clang_args}   )
```
[the remark this answers is not in the document]
Comment (nashif) [reply cmake/bintools/oneApi/target.cmake:8]: there is `/opt/intel/oneapi/intelpython/latest/bin/llvm-objdump`

File: doc/getting_started/toolchain_3rd_party_x_compilers.rst [hunk 4]
```diff
@@ -50,6 +50,18 @@ GNU ARM Embedded
...
+Intel oneApi Toolkits
+*********************
+
+#. Download `Intel oneAPI Base Toolkit
+   <https://software.intel.com/content/www/us/en/develop/tools/oneapi/all-toolkits.html>`_
+
+#. Assuming the toolkit is installed in ``/opt/intel/oneApi``, set environment
+   using::
+
+        source  /opt/intel/oneapi/setvars.sh
```
[the remark this answers is not in the document]
Comment (nashif) [reply doc/getting_started/toolchain_3rd_party_x_compilers.rst:62]: The setvars.sh does lots of things, doing this manually and listing them all should not be something we want to tinker with IMO

File: cmake/bintools/oneApi/target.cmake [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+find_program(CMAKE_RANLIB  llvm-ranlib  ${find_program_clang_args}   )
+find_program(CMAKE_OBJCOPY objcopy      ${find_program_binutils_args})
```
[the remark this answers is not in the document]
Comment (nashif) [reply cmake/bintools/oneApi/target.cmake]: we have it in /opt/intel/oneapi/intelpython/latest/bin/llvm-objcopy

[the remark this answers is not in the document]
Comment (nashif) [reply cmake/bintools/oneApi/target.cmake]: will change to llvm-objcopy, did not see any issues with gap-fill and x86 yet.

File: cmake/bintools/oneApi/target.cmake (hunk 2 above)
Comment (nashif) [reply cmake/bintools/oneApi/target.cmake:15]: now finding CMAKE_STRIP which is found as `/opt/intel/oneapi/intelpython/latest/bin/llvm-strip`

File: cmake/bintools/oneApi/target.cmake (hunk 3 above)
[the remark this answers is not in the document]
Comment (tejlmand) [reply cmake/bintools/oneApi/target.cmake:10]: Hmm, guess i'm missing something in my installation.
I disabled the Python, as it's almost 10Gb, and I just wanted the compiler.

Installing the `Intel distribution for Python` package helped.
10 Gb just for objdump and strip :laughing:

[the remark this answers is not in the document]
Comment (tejlmand) [reply cmake/bintools/oneApi/target.cmake:10]: would never have guessed those would be part of Python and not compiler if it were not for your link.

File: cmake/bintools/oneApi/target.cmake (hunk 5 above)
[the remark this answers is not in the document]
Comment (tejlmand) [reply cmake/bintools/oneApi/target.cmake]: I see this:
```
/opt/oneapi/intelpython/latest/bin/llvm-objcopy --gap-fill 0xff --output-target=binary --remove-section=.comment --remove-section=COMMON --remove-section=.eh_frame zephyr.elf zephyr.bin

/opt/oneapi/intelpython/latest/bin/llvm-objcopy: error: unknown argument '--gap-fill'
```

which is identical to earlier experience with vanilla llvm.
But gap fill is just a flash wearing improvement of filling gaps in flash with 0xFF, that is don't write anything compared to actually write 0x00.
I think it's acceptable to disable gap filling for toolchains not supporting this.

[the remark this answers is not in the document]
Comment (tejlmand) [reply cmake/bintools/oneApi/target.cmake]: Of course, one can build using:
```
 -DCONFIG_BUILD_NO_GAP_FILL=y
```
which points back to the fact we need to get proper toolchain / Kconfig structure in place for toolchain specific features.

File: cmake/linker/lld/target.cmake [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    # Note that the IMPLICIT_DEPENDS option is currently supported only
+    # for Makefile generators and will be ignored by other generators.
+    set(linker_script_dep IMPLICIT_DEPENDS C ${LINKER_SCRIPT})
+  elseif(CMAKE_GENERATOR STREQUAL "Ninja")
+    # Using DEPFILE with other generators than Ninja is an error.
+    set(linker_script_dep DEPFILE ${PROJECT_BINARY_DIR}/${linker_script_gen}.dep)
+  else()
+    # TODO: How would the linker script dependencies work for non-linker
+    # script generators.
+    message(STATUS "Warning; this generator is not well supported. The
```
Comment (tejlmand) [inline cmake/linker/lld/target.cmake:23]: Should this warning be a warning ?
```suggestion
    message(WARNING "This generator is not well supported. The
```

File: cmake/bintools/oneApi/target.cmake (hunk 5 above)
[the remark this answers is not in the document]
Comment (nashif) [reply cmake/bintools/oneApi/target.cmake]: r

> I see this:
> 
> ```
> /opt/oneapi/intelpython/latest/bin/llvm-objcopy --gap-fill 0xff --output-target=binary --remove-section=.comment --remove-section=COMMON --remove-section=.eh_frame zephyr.elf zephyr.bin
> 
> /opt/oneapi/intelpython/latest/bin/llvm-objcopy: error: unknown argument '--gap-fill'
> ```
> 
> which is identical to earlier experience with vanilla llvm.

where do you see this? build what code and what platform?

[the remark this answers is not in the document]
Comment (tejlmand) [reply cmake/bintools/oneApi/target.cmake]: This happens when building for `up_squared_32`:
```
$ cmake -GNinja -DBOARD=up_squared_32 -Bbuild samples/hello_world/
....
-- Generating done
-- Build files have been written to: /projects/github/zephyr_ws/zephyr/build
$ ninja -Cbuild
ninja: Entering directory `build'
...
/opt/oneapi/intelpython/latest/bin/llvm-objcopy: error: unknown argument '--gap-fill'
ninja: build stopped: subcommand failed.
```

Build is working when using `qemu_x86`.

**Comment by tejlmand** [comment]
When using `source <path>/<to>/oneApi/setvars.sh` the system Python settings get overruled with somithing that doesn't match my current installation, causing `menuconfig` to fail.

```
$ ninja -Cbuild menuconfig
ninja: Entering directory `build'
[0/1] cd /projects/github/ncs/zephyr/build/zephyr/kconfig && /opt/cmake-3.17.1-Linux-x86_64/bin/cmake -E env ZEPHYR_BASE=/projects/github/ncs/zephyr ZEPHYR_T...ephyr/dts/bindings /opt/oneapi/intelpython/latest/bin/python3.7 /projects/github/ncs/zephyr/scripts/kconfig/menuconfig.py /projects/github/ncs/zephyr/Kconfi
Traceback (most recent call last):
  File "/projects/github/ncs/zephyr/scripts/kconfig/menuconfig.py", line 194, in <module>
    import curses
  File "/opt/oneapi/intelpython/latest/lib/python3.7/curses/__init__.py", line 13, in <module>
    from _curses import *
ModuleNotFoundError: No module named '_curses'
```

Notice: **File "/opt/oneapi/intelpython/latest/lib/python3.7/curses/__init__.py"**.

Workaround, tell CMake to use specific Python, like:
```
cmake ....  -DPYTHON_PREFER=/usr/bin/python3.8 
```
and menuconfig works again.
Unsure what else could potentially be broken with the python from `oneApi` being in use.

File: cmake/bintools/oneApi/target.cmake (hunk 5 above)
[the remark this answers is not in the document]
Comment (nashif) [reply cmake/bintools/oneApi/target.cmake]: aha, was testing with qemu. How do you disable this per toolchain and avoid setting CONFIG_BUILD_NO_GAP_FILL for each board?

[the remark this answers is not in the document]
Comment (tejlmand) [reply cmake/bintools/oneApi/target.cmake]: > aha, was testing with qemu. How do you disable this per toolchain and avoid setting CONFIG_BUILD_NO_GAP_FILL for each board?

we don't at the moment :disappointed: 
Hence the reason for this comment earlier:
> which points back to the fact we need to get proper toolchain / Kconfig structure in place for toolchain specific features.

So we can take it in, and I'll look into cleanup of the Kconfig / Toolchain.

**Comment by nashif** [comment]
> When using `source <path>/<to>/oneApi/setvars.sh` the system Python settings get overruled with somithing that doesn't match my current installation, causing `menuconfig` to fail

yes i know, but this is going to be done independently, i.e. get away from sourcing the file and instead pulling only what we need. I will mention this in the docs.

File: cmake/bintools/oneApi/target.cmake (hunk 5 above)
[the remark this answers is not in the document]
Comment (nashif) [reply cmake/bintools/oneApi/target.cmake]: ok, addressed in the boards for now.

File: cmake/compiler/icx/generic.cmake [hunk 7]
```diff
@@ -0,0 +1,10 @@
+# SPDX-License-Identifier: Apache-2.0
+
+if(DEFINED TOOLCHAIN_HOME)
+  set(find_program_icx_args PATHS ${TOOLCHAIN_HOME} NO_DEFAULT_PATH)
+endif()
+
+find_program(CMAKE_C_COMPILER icx ${find_program_icx_args})
+
+find_program(CMAKE_DTS_PREPROCESSOR gcc)
```
Comment (ceolin) [inline cmake/compiler/icx/generic.cmake:9]: can't oneapi be used instead of host GCC ?

Comment (nashif) [reply cmake/compiler/icx/generic.cmake:9]: good point, just tried, and it actually works with icx

**Comment by nashif** [comment]
@tejlmand  I think I have addressed all the feedback

**Review by tejlmand — APPROVED** [review]

**Pull request merged**